### PR TITLE
SR-10555 foreignCopyUTF8 should do bulk access

### PR DESCRIPTION
**Pull request by Catfish-Man**
Fixes https://bugs.swift.org/browse/SR-10555 and rdar://50220164

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/StringBridge.swift [hunk 1]
```diff
@@ -69,6 +69,24 @@ internal func _cocoaStringSubscript(
...
+  return _swift_stdlib_CFStringGetBytes(
+    target,
+    _swift_shims_CFRange(location: 0, length: len),
+    kCFStringEncodingUTF8,
+    0,
+    0,
+    ptr,
+    bufPtr.count,
+    nil)
+}
```
Comment (milseman) [inline stdlib/public/core/StringBridge.swift]: What happens if it doesn't fit? Original code returned nil so we can issue the right precondition.

**Review by milseman — COMMENTED** [review]
Want to make sure we trigger the precondition failure. Might be worth adding a test case to StringTraps.swift for this.

File: stdlib/public/core/StringBridge.swift (hunk 1 above)
Comment (Catfish-Man) [reply stdlib/public/core/StringBridge.swift]: Ah yeah, need to replicate that. Will fix

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Comment by Catfish-Man** [comment]
Oops, forgot to make that conditional for ObjC

**Review by milseman — APPROVED** [review]
LGTM, let's see if it's covered by a benchmark

**Comment by milseman** [comment]
https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx-smoke-perf/902/

**Comment by Catfish-Man** [comment]
Alas, will have to write some new benchmarks

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
2.6x win, that’s more like it!

**Comment by milseman** [comment]
This seems related:
```swift
[ RUN      ] StringUTF8Tests.Contiguous Access
stdout>>> abcd
stdout>>> abcdefghijklmnop
stdout>>> abcdéfghijk
stderr>>> Fatal error: invalid Collection: less than 'count' elements in collection: file /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/swift/stdlib/public/core/ContiguousArrayBuffer.swift, line 643
stderr>>> Current stack trace:
stderr>>> 0    libswiftCore.dylib                 0x0000000109dfb020 _swift_stdlib_reportFatalErrorInFile + 116
stderr>>> 1    libswiftCore.dylib                 0x0000000109d14b80 specialized closure #1 in closure #1 in closure #1 in _fatalErrorMessage(_:_:file:line:flags:) + 304
stderr>>> 2    libswiftCore.dylib                 0x0000000109d14dc0 specialized closure #1 in closure #1 in _fatalErrorMessage(_:_:file:line:flags:) + 110
stderr>>> 3    libswiftCore.dylib                 0x0000000109cd2ff0 specialized closure #1 in _fatalErrorMessage(_:_:file:line:flags:) + 124
stderr>>> 4    libswiftCore.dylib                 0x0000000109cd3420 specialized _fatalErrorMessage(_:_:file:line:flags:) + 643
stderr>>> 5    libswiftCore.dylib                 0x0000000109cda0a0 specialized _copyCollectionToContiguousArray<A>(_:) + 900
stderr>>> 6    libswiftCore.dylib                 0x0000000109bbdf50 _StringGuts._foreignGrow(_:) + 56
stderr>>> 7    libswiftCore.dylib                 0x0000000109bbe170 _StringGuts.prepareForAppendInPlace(otherUTF8Count:) + 1323
stderr>>> 8    libswiftCore.dylib                 0x0000000109bbebf0 _StringGuts.append(_:) + 324
stderr>>> 9    libswiftCore.dylib                 0x0000000109cdb050 specialized _StringGuts.append(_:) + 230
stderr>>> 10   libswiftCore.dylib                 0x0000000109a5d8c0 static String.+= infix(_:_:) + 21
stderr>>> 11   String                             0x00000001090eb250 String.makeNative() + 49
stderr>>> 12   String                             0x00000001090eb3e0 closure #1 in  + 3736
stderr>>> 13   libswiftStdlibUnittest.dylib       0x00000001096fa890 specialized TestSuite._runTest(name:parameter:) + 584
stderr>>> 14   libswiftStdlibUnittest.dylib       0x00000001096b8ef0 _childProcess() + 846
stderr>>> 15   String                             0x00000001090ead30 main + 667
stderr>>> 16   libdyld.dylib                      0x00007fff62cc9ed8 start + 1
stderr>>> CRASHED: SIGILL
the test crashed unexpectedly
[     FAIL ] StringUTF8Tests.Contiguous Access
StringUTF8Tests: Some tests failed, aborting
UXPASS: []
FAIL: ["Contiguous Access"]
SKIP: []
To debug, run:
$ /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/validation-test-macosx-x86_64/stdlib/Output/StringUTF8.swift.tmp/String --stdlib-unittest-in-process --stdlib-unittest-filter "Contiguous Access"
```

**Comment by Catfish-Man** [comment]
Yeah, definitely something wrong around the edges of this. Will figure out what and land it on Monday :)

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
Linux failure is a known unrelated hang

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test linux platform

**Comment by Catfish-Man** [comment]
@swift-ci please test linux platform

**Comment by Catfish-Man** [comment]
oops. Guess those both worked >.<

**Comment by milseman** [comment]
What was the bug and why did this fix it? Also, can you test and make sure it gives the same result if the foreign string has unpaired surrogates?

**Pull request merged**